tephra-core/src/main/java/co/cask/tephra/persist/CommitMarkerCodec.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.tephra.persist;
+
+import co.cask.tephra.TxConstants;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Charsets;
+import com.google.common.primitives.Bytes;
+import com.google.common.primitives.Ints;
+import org.apache.hadoop.io.DataOutputBuffer;
+import org.apache.hadoop.io.SequenceFile;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+/**
+ * Class to read and write commit markers used in {@link AbstractHDFSLogReader} versions 2 and above.
+ */
+public class CommitMarkerCodec implements Closeable {
+  private static final byte[] KEY_BYTES = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED.getBytes(Charsets.UTF_8);
+  private final DataOutputBuffer rawKey;
+  private final ByteArrayOutputStream value;
+  private final DataOutputStream valueDataStream;
+  private SequenceFile.ValueBytes valueBytes;
+
+  public CommitMarkerCodec() {
+    this.rawKey = new DataOutputBuffer();
+    this.value = new ByteArrayOutputStream(Integer.SIZE);
+    this.valueDataStream = new DataOutputStream(value);
+  }
+
+  @Override
+  public void close() throws IOException {
+    valueDataStream.close();
+    value.close();
+    rawKey.close();
+  }
+
+  public int readMarker(SequenceFile.Reader reader) throws IOException {
+    if (valueBytes == null) {
+      valueBytes = reader.createValueBytes();
+    }
+    rawKey.reset();
+    value.reset();
+
+    int status = reader.nextRaw(rawKey, valueBytes);
+
+    //Check if the marker key is valid
+    if (status != -1 && isMarkerValid()) {
+      valueBytes.writeUncompressedBytes(valueDataStream);
+      valueDataStream.flush();
+      return Ints.fromByteArray(value.toByteArray());
+    }
+
+    // EOF not reached but couldn't read marker, then log an error
+    if (status != -1) {
+      throw new IOException(String.format("Invalid key for num entries appended found %s, expected : %s",
+                new String(rawKey.getData()), TxConstants.TransactionLog.NUM_ENTRIES_APPENDED));
+    }
+    return -1;
```

Comment:
If we handle the EOF case first, then the other two conditions become simple. It will also make code easy to read. What do you say?